File: apps/backend/src/services/metering.ts
import { db, usageEvents } from "@yomi/db"
import { eq } from "drizzle-orm"
import { user as userTable } from "../auth-schema.js"
import { effectivePlanForUser, hasBillablePlanAccess, getPlanConfig } from "../entitlements.js"
import { consumeCredits, getCreditSummary } from "./credit-ledger.js"
import { creditsForUsage, type BillableUsageKind } from "./credit-pricing.js"

// The four billable surfaces. Each maps to a credit cost (BillableUsageKind) and to
// the usageEvents.kind we persist for the dashboard breakdown.
export type ChargeKind = "chat" | "voice" | "analyze" | "bot_message" | "agent" | "composio_tool"

const CREDIT_KIND: Record<ChargeKind, BillableUsageKind> = {
  chat: "chat",
  voice: "voice",
  analyze: "analyze",
  bot_message: "bot_message",
  agent: "agent",
  composio_tool: "composio_tool",
}

// What we store on usageEvents.kind — chat/voice/agent become request_* to match
// the existing dashboard and transaction queries; bot_message/analyze/composio_tool
// pass through as-is.
const EVENT_KIND: Record<ChargeKind, string> = {
  chat: "request_chat",
  voice: "request_voice",
  analyze: "analyze",
  bot_message: "bot_message",
  agent: "request_agent",
  composio_tool: "composio_tool",
}

type MeteringUser = {
  id: string
  email?: string | null
  role?: string | null
  plan?: string | null
  subscriptionStatus?: string | null
  currentPeriodEnd?: Date | null
  trialEndDate?: Date | null
}

export type ChargeSuccess = {
  ok: true
  plan: string
  creditsRequired: number
  creditsCharged: number
  balance: number
  usageEventId: string | null
  paidBy: "credits"
}

export type ChargeFailure = {
  ok: false
  status: 402 | 403
  code: "subscription_inactive" | "subscription_required" | "credits_exhausted"
  message: string
  plan: string
}

export type ChargeResult = ChargeSuccess | ChargeFailure

function nextMonthResetLabel(): string {
  const now = new Date()
  const reset = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() + 1, 1))
  return reset.toLocaleDateString("en-US", { month: "long", day: "numeric", timeZone: "UTC" })
}

// Explore renews on the user's own trialEndDate (the explore-renewal cron tops
// credits back up there), not a calendar month boundary like paid plans.
function exploreRenewalLabel(user: MeteringUser): string {
  if (!user.trialEndDate) return "next month"
  return user.trialEndDate.toLocaleDateString("en-US", {
    month: "long",
    day: "numeric",
    timeZone: "UTC",
  })
}

// Single chokepoint for charging a billable action. Credits are the only gate: an
// active plan is required, and the action proceeds only when the credit balance
// covers the cost — then exactly one usage event is recorded and the credits are
// consumed. Callers must not do paid work before this returns ok.
export async function chargeUsage(input: {
  user: MeteringUser
  kind: ChargeKind
  durationSeconds?: number
  // Billed units for per-unit kinds (Composio tool calls in a turn). Defaults to 1.
  units?: number
  metadata?: Record<string, unknown>
}): Promise<ChargeResult> {
  const { user, kind } = input
  const plan = effectivePlanForUser(user)

  // Plan must be active (explore trial not expired, or paid sub active / in grace).
  if (!hasBillablePlanAccess(user)) {
    const status = user.subscriptionStatus ?? "inactive"
    const message =
      status === "past_due"
        ? "Payment didn't go through — Yomi is paused. Update your payment method in the dashboard."
        : status === "inactive" && plan === "explore"
          ? "Your free credits are renewing — try again in a moment, or upgrade to Pro or Max to skip the wait."
          : "Subscription isn't active. Head to the dashboard to sort it out."
    return { ok: false, status: 402, code: "subscription_inactive", message, plan }
  }

  const creditsRequired = creditsForUsage(CREDIT_KIND[kind], {
    durationSeconds: input.durationSeconds,
    units: input.units,
  })
  const summary = await getCreditSummary(user.id)

  // Out of credits — credits are the only gate, so block here.
  if (summary.balance < creditsRequired) {
    if (plan === "explore") {
      return {
        ok: false,
        status: 402,
        code: "subscription_required",
        message: `You're out of free credits for this month. They renew ${exploreRenewalLabel(user)}, or upgrade to Pro or Max for more right now.`,
        plan,
      }
    }
    return {
      ok: false,
      status: 402,
      code: "credits_exhausted",
      message: `You're out of credits. Buy a credit pack to continue. Resets ${nextMonthResetLabel()}.`,
      plan,
    }
  }

  // Record the usage event, then consume credits against it.
  const [event] = await db
    .insert(usageEvents)
    .values({
      userId: user.id,
      kind: EVENT_KIND[kind],
      inputTokens: 0,
      outputTokens: 0,
      costCents: input.durationSeconds ?? 0,
      creditsCharged: 0,
      status: "done",
      metadata: { reserveKind: kind, ...(input.metadata ?? {}) },
    })
    .returning({ id: usageEvents.id })

  if (!event) {
    // Could not record the event — fail closed rather than do unmetered work.
    console.error(`[metering] usage event insert returned no row user=${user.id} kind=${kind}`)
    return {
      ok: false,
      status: 402,
      code: "credits_exhausted",
      message: "Couldn't record usage. Please try again.",
      plan,
    }
  }

  const debit = await consumeCredits({
    userId: user.id,
    amount: creditsRequired,
    usageEventId: event.id,
    idempotencyKey: `usage:${event.id}:consume`,
    reason: `${kind} usage`,
    metadata: { kind, ...(input.metadata ?? {}) },
  })

  if (!debit.ok) {
    // Balance moved under us between the check and the debit (race). Block.
    return {
      ok: false,
      status: 402,
      code: plan === "explore" ? "subscription_required" : "credits_exhausted",
      message:
        plan === "explore"
          ? `You're out of free credits for this month. They renew ${exploreRenewalLabel(user)}, or upgrade to Pro or Max for more right now.`
          : `You're out of credits. Buy a credit pack to continue. Resets ${nextMonthResetLabel()}.`,
      plan,
    }
  }

  await db
    .update(usageEvents)
    .set({ creditsCharged: debit.charged })
    .where(eq(usageEvents.id, event.id))

  return {
    ok: true,
    plan,
    creditsRequired,
    creditsCharged: debit.charged,
    balance: debit.balance,
    usageEventId: event.id,
    paidBy: "credits",
  }
}

// Loads the plan/subscription fields chargeUsage needs, for callers that hold only
// a userId (e.g. the approval-replay path metering a Composio write).
export async function loadMeteringUser(userId: string): Promise<MeteringUser | null> {
  const [row] = await db
    .select({
      id: userTable.id,
      email: userTable.email,
      role: userTable.role,
      plan: userTable.plan,
      subscriptionStatus: userTable.subscriptionStatus,
      trialEndDate: userTable.trialEndDate,
      currentPeriodEnd: userTable.currentPeriodEnd,
    })
    .from(userTable)
    .where(eq(userTable.id, userId))
    .limit(1)
  return row ?? null
}

// Low-credit nudge: surfaced once remaining credits drop below 20% of the plan's
// monthly allotment. Returns null when there's nothing to warn about.
export function lowCreditWarning(user: MeteringUser, balance: number): string | null {
  const included = getPlanConfig(user).includedCredits
  if (included <= 0) return null
  if (balance > included * 0.2) return null
  return `You have ${balance} credits left.`
}

File: apps/backend/src/yomi/services/metering.py
"""Single credit-charging chokepoint.

Port of apps/backend/src/services/metering.ts. Credits are the only gate: an
active plan is required and the action proceeds only once the balance covers the
cost, then exactly one usage event is recorded and credits consumed.
"""

from __future__ import annotations

import logging
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Literal, TypedDict

from sqlalchemy import select, update
from sqlalchemy.dialects.postgresql import insert as pg_insert
from sqlalchemy.ext.asyncio import AsyncSession

from yomi.db.models_app import UsageEvent
from yomi.db.models_auth import User

from .credit_ledger import consume_credits, expire_credits, get_credit_summary
from .credit_pricing import UsagePricingInput, credits_for_usage
from .entitlements import (
    effective_plan_for_user,
    get_plan_config,
    has_billable_plan_access,
)

logger = logging.getLogger(__name__)

ChargeKind = Literal["chat", "voice", "analyze", "bot_message", "agent", "composio_tool"]

CREDIT_KIND: dict[ChargeKind, str] = {
    "chat": "chat",
    "voice": "voice",
    "analyze": "analyze",
    "bot_message": "bot_message",
    "agent": "agent",
    "composio_tool": "composio_tool",
}

# What we store on usage_events.kind to match existing dashboard/queries. Only
# chat/voice/agent become request_*; the rest pass through as-is.
EVENT_KIND: dict[ChargeKind, str] = {
    "chat": "request_chat",
    "voice": "request_voice",
    "analyze": "analyze",
    "bot_message": "bot_message",
    "agent": "request_agent",
    "composio_tool": "composio_tool",
}


class MeteringUser(TypedDict, total=False):
    id: str
    email: str | None
    role: str | None
    plan: str | None
    subscription_status: str | None
    current_period_end: datetime | None
    trial_end_date: datetime | None


@dataclass
class ChargeSuccess:
    ok: Literal[True] = True
    plan: str = ""
    credits_required: int = 0
    credits_charged: int = 0
    balance: int = 0
    usage_event_id: str | None = None
    paid_by: str = "credits"


@dataclass
class ChargeFailure:
    ok: Literal[False] = False
    status: int = 402
    code: str = "subscription_inactive"
    message: str = ""
    plan: str = ""


@dataclass
class ChargeInput:
    user: MeteringUser
    kind: ChargeKind
    duration_seconds: int | None = None
    units: int | None = None
    metadata: Mapping[str, object] | None = None


def _month_day(dt: datetime) -> str:
    return dt.strftime("%B").rstrip() + " " + str(dt.day)


def next_month_reset_label() -> str:
    now = datetime.now(UTC)
    reset = (now.replace(day=28) + timedelta(days=4)).replace(day=1)
    return _month_day(reset)


def explore_renewal_label(user: MeteringUser) -> str:
    trial_end = user.get("trial_end_date")
    if not trial_end:
        return "next month"
    return _month_day(trial_end)


async def charge_usage(session: AsyncSession, input_: ChargeInput) -> ChargeSuccess | ChargeFailure:
    user = input_.user
    kind = input_.kind
    plan = effective_plan_for_user(user)

    if not has_billable_plan_access(user):
        status = user.get("subscription_status") or "inactive"
        if status == "past_due":
            message = (
                "Payment didn't go through — Yomi is paused. Update your payment method "
                "in the dashboard."
            )
        elif status == "inactive" and plan == "explore":
            message = (
                "Your free credits are renewing — try again in a moment, or upgrade to "
                "Pro or Max to skip the wait."
            )
        else:
            message = "Subscription isn't active. Head to the dashboard to sort it out."
        return ChargeFailure(status=402, code="subscription_inactive", message=message, plan=plan)

    try:
        await expire_credits(session, user_id=user.get("id"))
    except Exception as exc:  # best-effort; never block charging on the sweep
        logger.error("[chargeUsage] expireCredits failed: %s %s", user.get("id"), exc)

    credits_required = credits_for_usage(
        CREDIT_KIND[kind],  # type: ignore[arg-type]
        UsagePricingInput(duration_seconds=input_.duration_seconds, units=input_.units),
    )
    summary = await get_credit_summary(session, user["id"])

    if summary.balance < credits_required:
        if plan == "explore":
            return ChargeFailure(
                ok=False,
                status=402,
                code="subscription_required",
                message=(
                    f"You're out of free credits for this month. They renew "
                    f"{explore_renewal_label(user)}, or upgrade to Pro or Max for more right now."
                ),
                plan=plan,
            )
        return ChargeFailure(
            ok=False,
            status=402,
            code="credits_exhausted",
            message=(
                f"You're out of credits. Buy a credit pack to continue. Resets "
                f"{next_month_reset_label()}."
            ),
            plan=plan,
        )

    event_id = (
        await session.execute(
            pg_insert(UsageEvent)
            .values(
                user_id=user["id"],
                kind=EVENT_KIND[kind],
                input_tokens=0,
                output_tokens=0,
                cost_cents=input_.duration_seconds or 0,
                credits_charged=0,
                status="done",
                metadata_={**(dict(input_.metadata or {})), "reserveKind": kind},
            )
            .returning(UsageEvent.id)
        )
    ).scalar_one_or_none()

    if event_id is None:
        logger.error(
            "[metering] usage event insert returned no row user=%s kind=%s", user.get("id"), kind
        )
        return ChargeFailure(
            status=402,
            code="credits_exhausted",
            message="Couldn't record usage. Please try again.",
            plan=plan,
        )

    debit = await consume_credits(
        session,
        user_id=user["id"],
        amount=credits_required,
        usage_event_id=str(event_id),
        idempotency_key=f"usage:{event_id}:consume",
        reason=f"{kind} usage",
        metadata={**(dict(input_.metadata or {})), "kind": kind},
    )

    if not debit.ok:
        code = "subscription_required" if plan == "explore" else "credits_exhausted"
        if plan == "explore":
            message = (
                f"You're out of free credits for this month. They renew "
                f"{explore_renewal_label(user)}, or upgrade to Pro or Max for more right now."
            )
        else:
            message = (
                f"You're out of credits. Buy a credit pack to continue. Resets "
                f"{next_month_reset_label()}."
            )
        return ChargeFailure(status=402, code=code, message=message, plan=plan)

    await session.execute(
        update(UsageEvent).where(UsageEvent.id == event_id).values(credits_charged=debit.charged)
    )

    return ChargeSuccess(
        plan=plan,
        credits_required=credits_required,
        credits_charged=debit.charged,
        balance=debit.balance,
        usage_event_id=str(event_id),
        paid_by="credits",
    )


async def load_metering_user(session: AsyncSession, user_id: str) -> MeteringUser | None:
    row = (
        await session.execute(
            select(User).where(User.id == user_id).limit(1)
        )
    ).scalar_one_or_none()
    if row is None:
        return None
    return MeteringUser(
        id=row.id,
        email=row.email,
        role=row.role,
        plan=row.plan,
        subscription_status=row.subscription_status,
        trial_end_date=row.trial_end_date,
        current_period_end=row.current_period_end,
    )


def low_credit_warning(user: MeteringUser, balance: int) -> str | None:
    included = get_plan_config(user)["includedCredits"]
    if included <= 0:
        return None
    if balance > included * 0.2:
        return None
    return f"You have {balance} credits left."